### ISSUE TITLE
Add test suite

### ISSUE THREAD

**Issue by SleeplessByte** [exercism/javascript-analyzer#9]
I prefer using something like [ava](https://github.com/avajs/ava) because it's nice for our needs (runs everything randomly and concurrently by default, and requires us to compile it first, which is the first step of the code).

We should:

- [x] test inputs (finding input files) and their flags
- [x] test outputs (that the output files are correctly created)
- [ ] for **each** exercise:
  - [ ] test at least 1 optimal solution
  - [ ] try to test hitting each path

Tracking progress:

- [ ] `two-fer`
  - [x] smoke (inline each status once)
  - [x] snapshot (0-20 fixture snapshot per status)
  - [ ] specific path tests
- [ ] `resistor-color`
  - [x] smoke (inline each status once)
  - [x] snapshot (0-20 fixture snapshot per status)
  - [ ] specific path tests
- [ ] `resistor-color-duo`
  - [x] smoke (inline each status once)
  - [x] snapshot (0-20 fixture snapshot per status)
  - [ ] specific path tests
- [ ] `gigasecond`
  - [x] smoke (inline each status once)
  - [x] snapshot (0-20 fixture snapshot per status)
  - [ ] specific path tests

**Comment by tejasbubane** [comment]
I have not used Typescript much, so excuse me if this is incorrect. But I feel we should stick to `Jest` for following reasons:

* Jest also runs everything randomly and concurrently by default.
* Having same tooling with the [JS track repo](https://github.com/exercism/javascript/) (apart from typescript) should be better. All decisions (test, eslint rules, code-conventions, etc.) apply to both repos - less bikeshedding.
* I hope compiling can be integrated in Jest as well (typescript-jest integration).

**Comment by SleeplessByte** [comment]
@tejasbubane yes, I'm fine with `jest` for the same reasons as I'm fine with `ava` (hence me saying "something like ava"). I think your reasoning makes sense. We _should_ also use `jest` in the `typescript` and `typescript-analyzer` repositories.

> I hope compiling can be integrated in Jest as well (typescript-jest integration).

As long as we don't use `ts-node` I'm good with that. I'm also okay with stripping types via `babel`, but that has to be a decision we make (to use babel instead of the `tsc` to to transpilation).

**Comment by tejasbubane** [comment]
Are fixtures used in any way right now?

**Comment by SleeplessByte** [comment]
> Are fixtures used in any way right now? If yes where? If not whats the plan with them?

Yes, these are the first 500 submissions to the exercise on the track. They can be used locally to run:
- batch-runner
- statistics

I use it to manually check if the right response is generated for an analyzer:
1. write the analyzer according to the optimal and minimal solution
1. inspect a few random solutions
1. make sure the analyzer works with that
1. run batch (all of them)
1. inspect at least 20 random exercises
1. repeat

The plan is to incorporate them in the tests to _automatically_ do that. Basically, as per above, a test could tag a sub-set with "exercise x, y and z should have output a, b, c"

**Comment by depsir** [comment]
Any reason why not using ts-jest?

From the  [documentation](https://basarat.gitbooks.io/typescript/docs/testing/jest.html)
> Install the TypeScript preprocessor for jest (ts-jest) which allows jest to transpile TypeScript on the fly and have source-map support built in.


I thing it should enable a quicker test workflow, allowing for example the `watch` option.
I tried it and the smoke test passes.

**Comment by SleeplessByte** [comment]
@depsir I see you've removed your comment. If you don't mind I'd still like to respond to it because it might help other wondering the same thing:

> [...] any reason why not using `ts-jest`? [...]

`ts-jest` was only a necessity when `ts-node` was still a _real solution_. It provided direct integration with compilation of typescript when running `jest` (and indeed also provide `--watch` support). 

Since babel 7 and TypeScript 3+, Microsoft has been working to help people **integrate** their TypeScript tooling with their JavaScript tooling, among other things enabling you to "compile" typescript via Babel (7+), essentially by stripping away the type information nodes. This allows us to rely on babel for **transpilation** ánd typescript for **type tooling** instead of: 
```
Source TypeScript > compilation (tsc/ts-node) > JavaScript > transpilation (babel) >  Output
```

Facebook no longer recommends `ts-jest` and I wouldn't be surprised if it will be deprecated in the (near) future.

**Comment by SleeplessByte** [comment]
There is now a document on [🌫 smoke tests](https://github.com/exercism/javascript-analyzer/blob/master/docs/smoke-tests.md) and [📸 snapshot tests](https://github.com/exercism/javascript-analyzer/blob/master/docs/snapshot-tests.md).

**Comment by SleeplessByte** [comment]
Fixed by #75
